### PR TITLE
Check seccomp enable and add unit test for seccomp/apparmor.

### PR DESCRIPTION
**Pull request by Random-Liu**
This PR:
1) Use `runc/libcontainer/seccomp` to check whether seccomp is enabled on OS.
2) Move apparmor and seccomp logic into helper functions.
3) Add unit test for apparmor and seccomp logic.

@mikebrow 
Signed-off-by: Lantao Liu <lantaol@google.com>

### CODE REVIEW

File: pkg/server/container_create.go [hunk 1]
```diff
@@ -768,6 +733,70 @@ func defaultRuntimeSpec() (*runtimespec.Spec, error) {
...
+		// use correct default profile (Eg. if not configured otherwise, the default is docker/default)
+		seccompProf = seccompDefaultProfile
+	}
+	if !seccompEnabled {
+		if seccompProf != "" && seccompProf != unconfinedProfile {
+			return nil, fmt.Errorf("seccomp is not supported")
+		}
+		return nil, nil
+	}
+	if privileged {
```
Comment (miaoyq) [inline pkg/server/container_create.go]: Maybe we could put the logic to the top of function.

If `privileged==true`, we could not care about the values of  `seccompEnabled` and `seccompProf`.

Comment (Random-Liu) [reply pkg/server/container_create.go]: Will do, make sense.

File: pkg/server/container_create.go [hunk 2]
```diff
@@ -201,59 +198,27 @@ func (c *criContainerdService) CreateContainer(ctx context.Context, r *runtime.C
...
-			if !strings.HasPrefix(appArmorProf, profileNamePrefix) {
-				return nil, fmt.Errorf("invalid apparmor profile %q", appArmorProf)
-			}
-			specOpts = append(specOpts, apparmor.WithProfile(strings.TrimPrefix(appArmorProf, profileNamePrefix)))
-		}
+
+	apparmorSpecOpts, err := generateApparmorSpecOpts(
+		securityContext.GetApparmorProfile(),
+		securityContext.GetPrivileged(),
+		runcapparmor.IsEnabled())
```
Comment (ijc) [inline pkg/server/container_create.go]: Perhaps the caller should call this itself?

File: pkg/server/container_create.go [hunk 3]
```diff
@@ -201,59 +198,27 @@ func (c *criContainerdService) CreateContainer(ctx context.Context, r *runtime.C
...
-			// Require and Trim default profile name prefix
-			if !strings.HasPrefix(seccompProf, profileNamePrefix) {
-				return nil, fmt.Errorf("invalid seccomp profile %q", seccompProf)
-			}
-			specOpts = append(specOpts, seccomp.WithProfile(strings.TrimPrefix(seccompProf, profileNamePrefix)))
-		}
+	seccompSpecOpts, err := generateSeccompSpecOpts(
+		securityContext.GetSeccompProfilePath(),
+		securityContext.GetPrivileged(),
+		runcseccomp.IsEnabled())
```
Comment (ijc) [inline pkg/server/container_create.go]: Perhaps the helper function should call `runccseccomp.IsEnbabled()` itself?

File: pkg/server/container_create.go (hunk 2 above)
Comment (Random-Liu) [reply pkg/server/container_create.go]: This is mainly for unit test. And actually, we don't need to call this every time, it does include some syscall.

I'll add global shared fields, and initialize those fields during startup.

File: pkg/server/container_create.go (hunk 1 above)
Comment (Random-Liu) [reply pkg/server/container_create.go]: Actually, CRI doesn't define the expected behavior clearly. We are not following docker's behavior.
https://github.com/moby/moby/blob/master/daemon/oci_linux.go#L751
https://github.com/moby/moby/blob/8af4db6f002ac907b6ef8610b237879dfcaa5b7a/daemon/seccomp_linux.go#L20

File: pkg/server/container_create.go [hunk 4]
```diff
@@ -201,59 +198,27 @@ func (c *criContainerdService) CreateContainer(ctx context.Context, r *runtime.C
...
-			}
-		default:
-			// Require and Trim default profile name prefix
-			if !strings.HasPrefix(appArmorProf, profileNamePrefix) {
-				return nil, fmt.Errorf("invalid apparmor profile %q", appArmorProf)
-			}
-			specOpts = append(specOpts, apparmor.WithProfile(strings.TrimPrefix(appArmorProf, profileNamePrefix)))
-		}
+
+	apparmorSpecOpts, err := generateApparmorSpecOpts(
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/server/container_create.go]: > thx for moving these into helpers that was next up on my list of things to do

Just get bored during the weekend, and addressed some TODOs. :)

[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/server/container_create.go]: > but it makes it really hard to review when also changing the logic at the same time

Sorry, messed up my commits, and have to combined them into one at last. :(

File: pkg/server/container_create.go [hunk 5]
```diff
@@ -201,59 +198,27 @@ func (c *criContainerdService) CreateContainer(ctx context.Context, r *runtime.C
...
+		securityContext.GetApparmorProfile(),
+		securityContext.GetPrivileged(),
+		runcapparmor.IsEnabled())
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate apparmor spec opts: %v", err)
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/server/container_create.go]: We prefixed all the errors in `CreateContainer`, and the error returned by the helper is short. I think it's fine to prefix here. :)

File: pkg/server/container_create.go [hunk 6]
```diff
@@ -201,59 +198,27 @@ func (c *criContainerdService) CreateContainer(ctx context.Context, r *runtime.C
...
-	if seccompProf == runtimeDefault || seccompProf == dockerDefault {
-		// use correct default profile (Eg. if not configured otherwise, the default is docker/default for containers)
-		seccompProf = seccompDefaultContainerProfile
-	}
-	// Unset the seccomp profile, if seccomp is not enabled, unconfined, unset, or the security context is privileged
-	if !seccompEnabled ||
-		seccompProf == unconfinedProfile ||
-		seccompProf == "" ||
-		config.GetLinux().GetSecurityContext().GetPrivileged() {
-		spec.Linux.Seccomp = nil
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/server/container_create.go]: Make sense. Let me see how to add it back.

File: pkg/server/container_create.go (hunk 3 above)
Comment (Random-Liu) [reply pkg/server/container_create.go]: ditto.

File: pkg/server/container_create.go (hunk 1 above)
Comment (mikebrow) [reply pkg/server/container_create.go]: Well there ya go.. in the absence of a defined spec... use the default behavior of the prior code

Comment (Random-Liu) [reply pkg/server/container_create.go]: > If privileged==true, we could not care about the values of seccompEnabled and seccompProf.

@miaoyq Done

File: pkg/server/container_create.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/server/container_create.go]: I updated our `defaultRuntimeSpec()` function to make sure no default apparmor/seccomp is specified by default.

File: pkg/server/container_create.go (hunk 2 above)
Comment (Random-Liu) [reply pkg/server/container_create.go]: Done.

**Comment by Random-Liu** [comment]
@mikebrow Addressed comments.

**Comment by Random-Liu** [comment]
@mikebrow Thanks for reviewing! :)

**Comment by Random-Liu** [comment]
Apply LGTM based on https://github.com/kubernetes-incubator/cri-containerd/pull/281#pullrequestreview-65068488

**Pull request merged**